File: Sources/Ably/ARTQueuedDealloc.m
#import "ARTQueuedDealloc.h"

@implementation ARTQueuedDealloc {
    id _ref;
    dispatch_queue_t _queue;
}

- (instancetype)init:(id)ref queue:(dispatch_queue_t)queue {
    self = [super init];
    if (self) {
        _ref = ref;
        _queue = queue;
    }
    return self;
}

- (void)dealloc {
    __block id ref = _ref;
    dispatch_async(_queue, ^{
        ref = nil;
    });
}

@end

File: Sources/AblySwift/ARTQueuedDealloc.swift
import Foundation

// swift-migration: original location ARTQueuedDealloc.h, line 5 and ARTQueuedDealloc.m, line 3
public class ARTQueuedDealloc: NSObject {
    
    private var _ref: AnyObject?
    private let _queue: DispatchQueue
    
    // swift-migration: original location ARTQueuedDealloc.m, line 8
    public init(ref: AnyObject, queue: DispatchQueue) {
        self._ref = ref
        self._queue = queue
        super.init()
    }
    
    // swift-migration: original location ARTQueuedDealloc.m, line 17
    deinit {
        let ref = _ref
        _queue.async {
            _ = ref // Keep reference alive until this block executes
        }
    }
}